File: app/src/main/java/com/example/smoothisland/XposedInit.java
package com.example.smoothisland;

import android.graphics.Outline;
import android.graphics.Path;
import android.view.View;
import de.robv.android.xposed.IXposedHookLoadPackage;
import de.robv.android.xposed.XC_MethodHook;
import de.robv.android.xposed.XposedBridge;
import de.robv.android.xposed.XposedHelpers;
import de.robv.android.xposed.callbacks.XC_LoadPackage.LoadPackageParam;

public class XposedInit implements IXposedHookLoadPackage {
    private static final String TAG = "SmoothIsland: ";
    private static final String SYSTEMUI_PKG = "com.android.systemui";

    @Override
    public void handleLoadPackage(LoadPackageParam lpparam) throws Throwable {
        if (!lpparam.packageName.equals(SYSTEMUI_PKG)) return;

        XposedBridge.log(TAG + "Manual Smooth Path Strategy active.");

        // 1. 接管 Outline 裁剪逻辑
        // 拦截 setRoundRect 并将其替换为我们手动计算的平滑 Path
        XposedHelpers.findAndHookMethod(Outline.class, "setRoundRect", 
            int.class, int.class, int.class, int.class, float.class, new XC_MethodHook() {
            @Override
            protected void beforeHookedMethod(MethodHookParam param) throws Throwable {
                Outline outline = (Outline) param.thisObject;
                int left = (int) param.args[0];
                int top = (int) param.args[1];
                int right = (int) param.args[2];
                int bottom = (int) param.args[3];
                float radius = (float) param.args[4];

                // 如果半径极小（接近直角），则不干预
                if (radius < 5) return;

                // 核心：生成手动计算的平滑圆角路径 (Squircle)
                Path squirclePath = createSmoothRectPath(left, top, right, bottom, radius);
                
                // 将 Outline 设置为 Path 模式
                // 这样系统会按照我们定义的贝塞尔曲线进行裁剪
                outline.setPath(squirclePath);
                
                // 阻止原有的 setRoundRect 执行，防止它覆盖我们的 Path
                param.setResult(null);
            }
        });

        // 2. 确保相关 View 开启了硬件加速裁剪
        XposedHelpers.findAndHookMethod(View.class, "onAttachedToWindow", new XC_MethodHook() {
            @Override
            protected void afterHookedMethod(MethodHookParam param) throws Throwable {
                View view = (View) param.thisObject;
                String name = view.getClass().getName();
                if (name.contains("Island") || name.contains("MiuiStatusIcon")) {
                    view.setClipToOutline(true);
                    // 禁用系统可能的原生平滑开关，避免冲突
                    try { 
                        XposedHelpers.callMethod(view, "setSmoothCornerEnabled", false); 
                    } catch (Throwable ignored) {}
                }
            }
        });
    }

    /**
     * 手动生成平滑圆角矩形路径 (Squircle 近似算法)
     * 使用三阶贝塞尔曲线，曲率比例 c 设为 0.72f 左右，能获得类似 HyperOS 的视觉饱满感
     */
    private Path createSmoothRectPath(int left, int top, int right, int bottom, float radius) {
        Path path = new Path();
        float w = right - left;
        float h = bottom - top;
        
        // 限制半径不超过最小边的一半
        float r = Math.min(radius, Math.min(w, h) / 2f);

        // 平滑系数：
        // 0.55228f 是标准圆弧
        // 0.72f 左右能达到明显的超椭圆平滑感
        float c = r * 0.72f; 

        path.reset();
        // 从左上角圆角终点顺时针绘制
        path.moveTo(left + r, top);
        
        // 顶部边
        path.lineTo(right - r, top);
        
        // 右上角平滑曲线
        path.cubicTo(right - r + c, top, right, top + r - c, right, top + r);
        
        // 右侧边
        path.lineTo(right, bottom - r);
        
        // 右下角平滑曲线
        path.cubicTo(right, bottom - r + c, right - r + c, bottom, right - r, bottom);
        
        // 底部边
        path.lineTo(left + r, bottom);
        
        // 左下角平滑曲线
        path.cubicTo(left + r - c, bottom, left, bottom - r + c, left, bottom - r);
        
        // 左侧边
        path.lineTo(left, top + r);
        
        // 左上角平滑曲线
        path.cubicTo(left, top + r - c, left + r - c, top, left + r, top);
        
        path.close();
        return path;
    }
}

File: app/src/main/java/com/example/smoothisland/XposedInit.kt
package com.example.smoothisland

import android.graphics.Outline
import android.graphics.Path
import android.graphics.Rect
import android.view.View
import androidx.graphics.shapes.RoundedPolygon
import androidx.graphics.shapes.pill
import androidx.graphics.shapes.toPath
import io.github.libxposed.api.XposedInterface
import io.github.libxposed.api.XposedModule
import io.github.libxposed.api.XposedModuleInterface.ModuleLoadedParam
import io.github.libxposed.api.XposedModuleInterface.PackageLoadedParam
import kotlin.math.abs

class XposedInit : XposedModule() {
    override fun onModuleLoaded(param: ModuleLoadedParam) {
        loadedProcessName = param.processName
    }

    override fun onPackageLoaded(param: PackageLoadedParam) {
        if (!param.isFirstPackage) return
        if (loadedProcessName != SYSTEMUI_PKG) return
        if (param.packageName != SYSTEMUI_PKG) return

        installHooks(param.defaultClassLoader)
    }

    private fun overrideOutlineIfCapsule(outline: Outline) {
        val bounds = Rect()
        if (outline.getRect(bounds) && bounds.height() > 10) {
            val height = bounds.height()
            if (abs(outline.radius - (height / 2.0f)) <= 1.5f) {
                val path = createSmoothCapsulePath(
                    bounds.left,
                    bounds.top,
                    bounds.right,
                    bounds.bottom
                )
                outline.setPath(path)
            }
        }
    }

    private fun createSmoothCapsulePath(
        left: Int,
        top: Int,
        right: Int,
        bottom: Int
    ): Path {
        val width = right - left
        val height = bottom - top
        val path = getBaseCapsulePath(width, height)
        path.offset(left + (width / 2.0f), top + (height / 2.0f))
        return path
    }

    @Synchronized
    private fun getBaseCapsulePath(width: Int, height: Int): Path {
        val smoothing = readSmoothing()
        val key = CapsuleShape(width, height, smoothing)
        val cachedPath = capsulePathCache[key]
        if (cachedPath != null) {
            return Path(cachedPath)
        }

        val generatedPath = RoundedPolygon.pill(
            width = width.toFloat(),
            height = height.toFloat(),
            smoothing = smoothing
        ).toPath()

        if (capsulePathCache.size >= MAX_PATH_CACHE_SIZE) {
            val oldestKey = capsulePathCache.entries.iterator().next().key
            capsulePathCache.remove(oldestKey)
        }
        capsulePathCache[key] = generatedPath
        return Path(generatedPath)
    }

    private fun readSmoothing(): Float {
        val rawValue = getSystemProperty(SMOOTHING_PROP, DEFAULT_SMOOTHING.toString())
        val parsedValue = rawValue.toFloatOrNull() ?: DEFAULT_SMOOTHING
        return parsedValue.coerceIn(MIN_SMOOTHING, MAX_SMOOTHING)
    }

    private fun getSystemProperty(key: String, defaultValue: String): String {
        return try {
            val systemProperties = Class.forName("android.os.SystemProperties")
            val getMethod = systemProperties.getMethod(
                "get",
                String::class.java,
                String::class.java
            )
            getMethod.invoke(null, key, defaultValue) as String
        } catch (_: Throwable) {
            defaultValue
        }
    }

    @Synchronized
    private fun installHooks(classLoader: ClassLoader) {
        if (hooksInstalled) return

        hookOutlineRoundRectForDynamicIsland()
        hookTargetOutlineProviders(classLoader)
        hooksInstalled = true
    }

    private fun hookTargetOutlineProviders(classLoader: ClassLoader) {
        TARGET_PROVIDER_CLASSES.forEach { providerClassName ->
            try {
                val targetProvider = Class.forName(providerClassName, false, classLoader)
                val getOutlineMethod = targetProvider.getDeclaredMethod(
                    "getOutline",
                    View::class.java,
                    Outline::class.java
                )

                hook(getOutlineMethod)
                    .setPriority(XposedInterface.PRIORITY_DEFAULT)
                    .intercept { chain ->
                        val result = chain.proceed()
                        val outline = chain.getArg(1) as Outline
                        overrideOutlineIfCapsule(outline)
                        result
                    }
            } catch (_: Throwable) {
            }
        }
    }

    private fun hookOutlineRoundRectForDynamicIsland() {
        val setRoundRectMethod = Outline::class.java.getDeclaredMethod(
            "setRoundRect",
            Int::class.javaPrimitiveType!!,
            Int::class.javaPrimitiveType!!,
            Int::class.javaPrimitiveType!!,
            Int::class.javaPrimitiveType!!,
            Float::class.javaPrimitiveType!!
        )

        hook(setRoundRectMethod)
            .setPriority(XposedInterface.PRIORITY_DEFAULT)
            .intercept { chain ->
                val left = chain.getArg(0) as Int
                val top = chain.getArg(1) as Int
                val right = chain.getArg(2) as Int
                val bottom = chain.getArg(3) as Int
                val radius = chain.getArg(4) as Float
                val height = bottom - top

                if (
                    height > 10 &&
                    abs(radius - (height / 2.0f)) <= 1.0f &&
                    isDynamicIslandOutlineCall()
                ) {
                    val path = createSmoothCapsulePath(left, top, right, bottom)
                    val outline = chain.thisObject as Outline
                    outline.setPath(path)
                    null
                } else {
                    chain.proceed()
                }
            }
    }

    private fun isDynamicIslandOutlineCall(): Boolean {
        var hasDynamicIslandCaller = false

        Thread.currentThread().stackTrace.forEach { frame ->
            val className = frame.className
            val lowerClassName = className.lowercase()

            if (EXCLUDED_OUTLINE_CALLERS.any { lowerClassName.contains(it) }) {
                return false
            }
            if (DYNAMIC_ISLAND_CALLERS.any { lowerClassName.contains(it) }) {
                hasDynamicIslandCaller = true
            }
        }

        return hasDynamicIslandCaller
    }

    private companion object {
        const val SYSTEMUI_PKG = "com.android.systemui"
        val TARGET_PROVIDER_CLASSES = listOf(
            "com.android.systemui.statusbar.notification.DynamicIslandWindowAnimController\$updateFakeViewOutline\$1",
            "com.android.systemui.statusbar.notification.mediaisland.MiuiIslandMediaViewHolder\$Companion\$create\$1\$1"
        )
        val DYNAMIC_ISLAND_CALLERS = listOf(
            "dynamicisland",
            "mediaisland"
        )
        val EXCLUDED_OUTLINE_CALLERS = listOf(
            "footerview",
            "footerviewbutton",
            "notificationstackscrolllayout",
            "notif_footer"
        )
        const val SMOOTHING_PROP = "persist.smoothisland.smoothing"
        const val DEFAULT_SMOOTHING = 0.8f
        const val MIN_SMOOTHING = 0.0f
        const val MAX_SMOOTHING = 1.0f
        const val MAX_PATH_CACHE_SIZE = 32
        val capsulePathCache = LinkedHashMap<CapsuleShape, Path>(MAX_PATH_CACHE_SIZE, 0.75f, true)
        @Volatile
        var hooksInstalled = false
        @Volatile
        var loadedProcessName: String? = null
    }

    private data class CapsuleShape(val width: Int, val height: Int, val smoothing: Float)
}
